### PR TITLE
232 create circuit components for fri verifier

### LINKED ISSUE #232: FRI cosets generator component

### PR DESCRIPTION
**Pull request by ayashunsky**
Implementation of the FRI cosets circuit component and its tests
Closes #232

### CODE REVIEW

**Review by Iluvmagick — CHANGES_REQUESTED** [review]

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -133,6 +133,7 @@ set(PLONK_TESTS_FILES
     "verifiers/pickles/verify_heterogenous_scalar"
     "verifiers/pickles/verify_heterogenous_base"
     "verifiers/pickles/scalar_details/evals_of_split_evals"
+    "verifiers/placeholder/fri_cosets"
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply test/CMakeLists.txt:136]: Not enough rights

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+#include <nil/blueprint/blueprint/plonk/circuit.hpp>
+#include <nil/blueprint/component.hpp>
+#include <nil/blueprint/manifest.hpp>
+
+namespace nil {
+    namespace blueprint {
+        namespace components {
+            // compute the number of lines if we need to place o object with pl object per line
+            template<typename T1, typename T2>
+            T1 lfit(T1 o, T2 pl) {
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp]: done

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+                 return o/pl + (o % pl > 0);
+            }
+
+	    // Uses parameters  n, total_bits, omega
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp]: done, I hope

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+                    std::uint32_t gates_amount() const override {
+                        return fri_cosets::gates_amount_internal(witness_amount,n,total_bits);
+                    }
+
+                    bool operator<(const component_gate_manifest *other) const override {
+                        std::size_t o_witness_amount = dynamic_cast<const gate_manifest_type*>(other)->witness_amount;
+                        std::size_t o_n = dynamic_cast<const gate_manifest_type*>(other)->n;
+                        std::size_t o_total_bits = dynamic_cast<const gate_manifest_type*>(other)->total_bits;
+
+                        std::size_t l = witness_amount / 6;       std::size_t o_l = o_witness_amount / 6;
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp]: Created a function for that

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+                    // Applying gate type 4 to lines sixb_rows -- (sixb_rows + lfit(remaining_bl2,bl2_line) - 2)
+                    for(std::size_t i = 0; i < lfit(remaining_bl2,bl2_line) - 1; i++) {
+                        assignment.enable_selector(selector_index, start_row_index + sixb_rows + i);
+                    }
+                }
+
+                std::size_t last_bl2 = remaining_bl2 % bl2_line; // number of 2-blocks in the last 2-block line
+                if (last_bl2 == 0) { last_bl2 = bl2_line; }
+                if (remaining_bl2 > 0) { // if 2-blocks don't all fit into transition row (gate type 5)
+                    cs.clear();
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp]: done

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+                    selector_index = bp.add_gate(six_block); // type 2 gate
+                    // Applying gate type 2 to lines 1--(sixb_rows - 2)
+                    for(std::size_t i = 1; i < sixb_rows - 1; i++) {
+                        assignment.enable_selector(selector_index, start_row_index + i);
+                    }
+                }
+
+                // The gate for the line where the 6-blocks end
+                std::size_t last = (last_l > 0)? last_l : l; // The number of the last 6-block in the row
+                cs = {six_block[0]};
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp]: See above, done

File: test/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Iluvmagick) [reply test/CMakeLists.txt:136]: Ok, added myself.

**Review by Iluvmagick — APPROVED** [review]

**Pull request merged**